sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py

```diff
@@ -246,18 +247,22 @@ def __next__(self):
     next = __next__  # for python2.7
 
     def _iter_next(self):
-        self._open()
-        if not self._message_iter:
-            self._message_iter = self._handler.receive_messages_iter()
-        uamqp_message = next(self._message_iter)
-        message = self._build_message(uamqp_message)
-        if (
-            self._auto_lock_renewer
-            and not self._session
-            and self._receive_mode != ServiceBusReceiveMode.RECEIVE_AND_DELETE
-        ):
-            self._auto_lock_renewer.register(self, message)
-        return message
+        try:
+            self._receive_context.set()
+            self._open()
+            if not self._message_iter:
+                self._message_iter = self._handler.receive_messages_iter()
+            uamqp_message = next(self._message_iter)
+            message = self._build_message(uamqp_message)
+            if (
+                self._auto_lock_renewer
+                and not self._session
+                and self._receive_mode != ServiceBusReceiveMode.RECEIVE_AND_DELETE
+            ):
+                self._auto_lock_renewer.register(self, message)
+            return message
+        finally:
+            self._receive_context.clear()
```

Comment:
If I recall correctly, we had a system in place to allow for iterator receiving and batch message receiving from the same receiver.... is that correct?
And if so - how well is that handled when they're both sharing the same threading.Event?

Reply:
both would use the same threading.Event.

execution in sequential order is fine as we're doing set and clear in sequential order..

however, execution in nested situation/parallel execution would be a problem.. okay, I think I need to change it to counter variable instead of this threading.Event..

thanks for raising your concern!

Reply:
Do we have any tests that use both iterator an batch at the same time?

Reply:
test added for iterator + iterator and iterator + batch